### ISSUE TITLE
Popular LLMs don't support echo + logprob for reliability score calc

### ISSUE THREAD

**Issue by eventlOwOp** [Li-Z-Q/DeepSolution#2]
Dear Author(s),
 
First of all, thank you for sharing this repository and the corresponding paper. It has been incredibly helpful for understanding and reproducing your work. I’ve encountered an issue while running the code and would greatly appreciate your guidance.
 
## Problem Description
When trying to calculate scores using the code in this repository, I noticed that mainstream large language models (including all OpenAI models and DeepSeek models, etc.) do not support the combination of  echo=True  and  logprob=True . Specifically:
 
All the models ignore the echo param.
 
This issue prevents the code from correctly calculating scores as intended.

Could you please suggest:
 
An alternative approach or parameter combination to achieve the same functionality?
 
Whether there are known workarounds or updates to address this issue?

**Comment by Li-Z-Q** [comment]
Thank you for your attention. 

In all experiments of this work, we deployed the model locally using vLLM, with `echo=True` and `logprob=True` being parameters supported by vLLM.

If you wish to use the APIs from OpenAI or DeepSeek, you will need to refer to their respective parameters. To the best of my knowledge, the corresponding parameter in OpenAI's API is `logprobs=True`, as shown in the figure. You may try setting it accordingly.

![Image](https://github.com/user-attachments/assets/4893b831-c3f3-4329-8d41-ad60ebef449f)

If you have further questions or need more discussion, please feel free to contact us at any time.

**Comment by eventlOwOp** [comment]
Dear Author,

Thank you very much for your response and for pointing out that OpenAI and DeepSeek provide the logprobs parameter. I appreciate your time and assistance.

While `logprobs=True` does return token-level probabilities for the completion, it does not include the logprob values for the prompt itself. Additionally, the `echo` parameter, which would allow the model to return the prompt along with its logprob values, is not supported by OpenAI, DeepSeek, or many other cloud-based LLM APIs.

This limitation seems to prevent the direct application of the reliability score calculation method described in the paper to cloud-based LLM APIs. Specifically, since the prompt’s logprob values are not accessible, the method cannot be implemented as-is.

Given that our application relies on cloud-based APIs😭, could you kindly suggest any alternative approaches to the reliability score calculation method that would be compatible with cloud-based LLM APIs? Or just it's impossible to use DeepSolution in this situation which we'll be sorry to hear because this project values greatly to our work😭.

Thank you again for your time and support. Your insights would be highly valuable in helping us adapt this work to our application.

**Comment by Li-Z-Q** [comment]
I am very sorry that my last response did not address your issue. Now I completely understand the problem you are encountering. Indeed, as you said, using just `logprobs=True` does not enable cloud-based APIs to return the overall probability value of the text.

For the situation you are facing, I suggest that you could make some modifications to our scoring method. Instead of using a logits-based scoring approach, you could directly have the LLM output a score. Specifically, you could concatenate the `solution `and `comment `and input them into the LLM, instructing it to generate a score from 0 to 10 for the solution . Similarly, you could concatenate the `old solution`, `comment`, and `new solution`, and input them into the LLM, instructing it to generate a score from 0 to 10 for the comment . The lines of code you would need to modify are lines 328 and 349 in `ours_framework.py`.

The reason we did not use such a text-based scoring method in our experiments is that we were concerned that the 7B model we used might not have the capability to perform such tasks effectively. However, models like OpenAI or Deepseek should have no problem scoring in this textual manner.

Lastly, if you do intend to use OpenAI or Deepseek to build our framework, we recommend that you appropriately reduce the depth of the tree. On one hand, this can save costs, and on the other hand, based on my intuition, allowing a highly capable model like OpenAI or Deepseek to iterate too many times might not yield better results.

I hope the above information is helpful to you. If you have any further questions or need more discussion, please feel free to contact us anytime.

**Comment by eventlOwOp** [comment]
Thank you so much for your detailed and insightful response! I truly appreciate the time and effort you took to address my concerns and provide a clear solution to the issue.

**Comment by eventlOwOp** [comment]
we've done some work to realize realtime tree visualizationi

but we've also removed some files to adapt it to our application, so it could be hard to merge.

[add realtime visualization; add openai api support; add RAG base class for HippoRAG;](https://github.com/eventlOwOp/DeepSolution/commit/f8436cfdfa8542f858a3adff0fa9fbbf3072dbfd)
